feat: store deployed forms by id and version tag

jfriedenstab opened the pull request:
## Description
Store the version tag as a new property for deployed forms and add the possibility to find a form by form id and version tag. This is a prerequisite to be able to properly resolve linked forms when the new `versionTag` binding type is used in a user task (see #21041).

## Related issues
closes #21037

koevskinikola commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/transform/FormResourceTransformer.java:
```diff
@@ -91,15 +92,14 @@ public Either<Failure, Void> writeRecords(
       final var failureMessage =
           String.format(
-              "Failed to parse formId from form JSON. '%s': %s",
+              "Failed to parse form JSON. '%s': %s",
```
🤔 I think this is a potential breaking change, as some users may parse these messages and react on them.

@korthout do you think changing this message is acceptable? Forms are a "relatively" new feature. Furthermore, we don't guarantee the content of error messages (afaik).

korthout replied:
@koevskinikola I get this concern, but so far we've never made an issue out of this AFAIK. If we were to take our promises about not breaking user space to this level, it would become very hard to further improve things. I.e. in order to change error messages we'd need some configuration option to use the "improved" error messages.

I'd argue that:
- the error type can be depended upon as they are documented in the gRPC API (and perhaps also in the REST API 🤔)
- the error message provide additional information, but are not documented and can thus not be depended upon

jfriedenstab replied:
Thanks! So nothing to be changed here :).

koevskinikola commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/mutable/MutableFormState.java:
```diff
@@ -67,4 +74,11 @@ public interface MutableFormState extends FormState {
    * @param record the record of the form that is deleted
    */
   void deleteFormInFormKeyByFormIdAndDeploymentKeyColumnFamily(FormRecord record);
+
+  /**
+   * Deletes a form from FORM_KEY_BY_FORM_ID_AND_VERSION_TAG column family
+   *
+   * @param record the record of the form that is deleted
+   */
+  void deleteFormInFormKeyByFormIdAndVersionTagColumnFamily(FormRecord record);
```
🔧 It would be nice if deletion-related code is in a separate commit for better PR readability (and future understanding of the changes).

jfriedenstab replied:
See https://github.com/camunda/camunda/pull/21494/commits/4a1c4f6bd7bb26a8c5e7b8d09dd5acca14c7ef8c.

koevskinikola commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/FormDeploymentTest.java:
🔧 A test that confirms that a new form deployment with the same version tag overwrites the mapping would be useful as well.

jfriedenstab replied:
I've added new tests with https://github.com/camunda/camunda/pull/21494/commits/93b24168e71c286a2af1787816841142c7be3629 and https://github.com/camunda/camunda/pull/21494/commits/eb13b71bd86d9b19fcc9d45be6357b995fe4fa46.

koevskinikola commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionTest.java:
🔧 It would be nice if deletion-related code is in a separate commit for better PR readability (and future understanding of the changes).

jfriedenstab replied:
See https://github.com/camunda/camunda/pull/21494/commits/9f058ff8703c17ebeb89200fa701bd08af077ed6.

koevskinikola commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/appliers/AbstractFormCreatedApplierTest.java:
```diff
@@ -123,6 +123,7 @@ Consumer<PersistedForm> isEqualToFormRecord(final FormRecord record) {
       assertThat(bufferAsArray(persistedForm.getChecksum())).isEqualTo(record.getChecksum());
       assertThat(bufferAsArray(persistedForm.getResource())).isEqualTo(record.getResource());
       assertThat(persistedForm.getDeploymentKey()).isEqualTo(record.getDeploymentKey());
+      assertThat(persistedForm.getVersionTag()).isEqualTo(record.getVersionTag());
```
❌ I think this change will require a new version of the applier, as (form deletion command) records from older Zeebe versions might be processed. 

@korthout just pinging you here to confirm this.

korthout replied:
👍 @koevskinikola This specific change is fine.

This change only ensures that two FormRecord objects are equivalent. Older `Form:CREATED` records will not have the version tag property set, and so it will be considered to contain the default `""` empty string. These should be taken into account when comparing them.

https://github.com/camunda/camunda/blob/078a73566bb5fbdffe20414090b3683f02ce9279/zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/deployment/FormRecord.java#L34

Note that the version tag is only stored in the new column family when the new v2 `From:CREATED` is appended. So I think that should be fine. 

❌ The only thing I noticed, is that we are still appending v1 of the `Form:CREATED` event in `FormResourceTransformer` and `DeploymentCreateProcessor`.

jfriedenstab replied:
> The only thing I noticed, is that we are still appending v1 of the Form:CREATED event in `FormResourceTransformer` and `DeploymentCreateProcessor`.

Hi @korthout,
Actually, I think version 2 is used, because it's the latest available one. There's also a test that checks the record version:

https://github.com/camunda/camunda/blob/008d8b9106088d19d4895936b6846254d019b1c9/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/FormDeploymentTest.java#L107

korthout replied:
Ah, shoot! Of course! I forgot about this 🙇 Thanks

koevskinikola commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbFormState.java:
```diff
@@ -170,6 +206,14 @@ public void deleteFormInFormKeyByFormIdAndDeploymentKeyColumnFamily(final FormRe
     formKeyByFormIdAndDeploymentKeyColumnFamily.deleteIfExists(tenantAwareFormIdAndDeploymentKey);
   }
 
+  @Override
+  public void deleteFormInFormKeyByFormIdAndVersionTagColumnFamily(final FormRecord record) {
+    tenantIdKey.wrapString(record.getTenantId());
+    dbFormId.wrapString(record.getFormId());
+    dbVersionTag.wrapString(record.getVersionTag());
+    formKeyByFormIdAndVersionTagColumnFamily.deleteIfExists(tenantAwareFormIdAndVersionTagKey);
+  }
+
```
🔧 It would be nicer to put this change in a separate commit, for better understanding of the changes in the future.

jfriedenstab replied:
I've extracted the deletion related changes into a separate [commit](https://github.com/camunda/camunda/pull/21494/commits/4a1c4f6bd7bb26a8c5e7b8d09dd5acca14c7ef8c).

koevskinikola reviewed (changes requested):
Nice job @jfriedenstab 🎉 

There are some nitpicks, minor suggestions, and feedback on how the deletion applier should be changed. I pinged @korthout to confirm my feedback just in case.

jfriedenstab commented on zeebe/exporters/elasticsearch-exporter/src/main/resources/zeebe-record-deployment-template.json:
I moved the changes to a separate commit.

jfriedenstab commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/appliers/FormDeletedApplierTest.java:
I extracted the deletion-related changes into a separate commit.

jfriedenstab commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/FormDeletedApplier.java:
I've extracted the changes into a separate commit.

Regarding:
> I think a V2 of this applier is also needed since (form delete) records from earlier Zeebe versions might be processed with this one.

I'm not exactly sure about that. Since the delete method in the `DbFormState` uses `deleteIfExists`, processing older form delete records with the new applier would probably have no impact?

By the way, I had a similar discussion with Nico here: https://github.com/camunda/camunda/pull/20498#discussion_r1691653526. So, I just added the deletion from the new column families to the existing event applier (for processes, decisions, and forms).

koevskinikola replied:
We had a similar discussion in the Decisions-related PR. I think the changes are okay as they are.

jfriedenstab commented on zeebe/protocol/src/main/java/io/camunda/zeebe/protocol/record/value/deployment/FormMetadataValue.java:
```diff
@@ -59,4 +59,9 @@ public interface FormMetadataValue extends RecordValue, TenantOwned {
+  /**
+   * @return the custom version tag of the form
+   */
+  String getVersionTag();
```
See https://github.com/camunda/camunda/pull/21494/commits/6574d34a3374e94579ca2c8c67ce876c83e1411c.

jfriedenstab commented on zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/deployment/FormRecord.java:
```diff
@@ -131,6 +134,16 @@ public FormRecord setDeploymentKey(final long deploymentKey) {
+    versionTagProp.setValue(versionTag);
+    return this;
+  }
+
```
See https://github.com/camunda/camunda/pull/21494/commits/6574d34a3374e94579ca2c8c67ce876c83e1411c.

jfriedenstab commented:
@koevskinikola I've addressed all your comments except [this one](https://github.com/camunda/camunda/pull/21494#discussion_r1728925446). Please have a look at my [response](https://github.com/camunda/camunda/pull/21494#discussion_r1730786656).

koevskinikola reviewed (approved):
🚀
